File: Leetcode/WaystoExpressanIntegerasSumofPowers.rs
// https://leetcode.com/problems/ways-to-express-an-integer-as-sum-of-powers

struct Solution {}

impl Solution {
    const MOD: usize = 1_000_000_000 + 7;

    pub fn number_of_ways(n: i32, x: i32) -> i32 {
        let (n, x) = (n as usize, x as usize);

        let mut nums = vec![];
        for i in 1..=n {
            let mut ii = 1;
            for _ in 0..x {
                ii *= i;
            }
            nums.push(ii);
        }

        let mut dp = vec![vec![0; nums.len() + 1]; n + 1];
        for j in 0..=nums.len() {
            dp[0][j] = 1;
        }
        for i in 1..=n {
            for j in 1..=nums.len() {
                dp[i][j] += dp[i][j - 1];
                dp[i][j] %= Solution::MOD;
                if i >= nums[j - 1] {
                    dp[i][j] += dp[i - nums[j - 1]][j - 1];
                    dp[i][j] %= Solution::MOD;
                }
            }
        }

        dp[n][nums.len()] as i32
    }
}

File: Leetcode/WaystoExpressanIntegerasSumofPowers.cpp
// https://leetcode.com/problems/ways-to-express-an-integer-as-sum-of-powers

class Solution {
   public:
    int numberOfWays(int n, int x) {
        using i64 = long long;
        const i64 MOD = 1000000000 + 7;

        vector<i64> nums;
        for (i64 i = 1; i <= n; ++i) {
            auto curi = i;
            for (i64 j = 1; j < x; ++j) {
                curi *= i;
            }
            nums.push_back(curi);
        }

        const auto m = nums.size();
        vector<vector<i64>> dp(n + 1, vector<i64>(m + 1, 0));

        dp[0][0] = 1;
        for (auto i = 1; i <= m; ++i) {
            const auto num = nums[i - 1];

            for (auto j = 0; j <= n; ++j) {
                dp[j][i] += dp[j][i - 1];
                dp[j][i] %= MOD;

                if (j + num <= n) {
                    dp[j + num][i] += dp[j][i - 1];
                    dp[j + num][i] %= MOD;
                }
            }
        }

        return dp[n][m];
    }
};